observablehq/framework
# Distinguish between URL paths and file paths at the type level

This is an attempt to improved Windows support. The idea here is that we can use TypeScript to tag some strings as being either `FilePath`s or `UrlPath`s. From there we can make a series of helpers around the Node filesystem and path functions that are stricter about which one they take.

I'm putting this up as a draft now to get some early feedback about this approach. I think there is some inherent boilerplate here to tag strings as the two types. I think there may also be some boilerplate in this PR that we could get rid of. Any comments or suggestions about how we could improve this are welcome!

As of now, I have all the tests passing on Windows locally.

Besides the path annotations, I've introduced some libraries to ease in cross-platform compatibility:
- `cross-env` - Handles setting environment variables inline in `package.json` scripts
- `rimraf` - A cross platform `rm`-like tool
- `cross-spawn` - For executing files more easily in a cross platform way.

I'm not sure how we can get `.sh` data loaders working, so for now I've simply disabled those tests on Windows. I think we can save that for another PR.

In order for our snapshot tests to work, I added a `.gitattributes` file that insists on posix-style line endings everywhere. I based that off of [these GitHub docs](https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings) and [this blogpost](https://adaptivepatchwork.com/2012/03/01/mind-the-end-of-your-line/).

---

# To Test

- Check out this branch
- Run `yarn install` to update packages
- Run `yarn observable create` to create a new project
- `cd` into the project directory
- Edit the project to use the branch instead of the npm published version of Framework:
  - `yarn remove @observablehq/framework`
  - `yarn add link:..`
- Use Framework with commands like `yarn dev`, `yarn build`, `yarn observable login` and `yarn deploy`.
